### PR TITLE
Fixes memory leak in messaging service

### LINKED ISSUE #6878: Memory leak in NettyMessagingService
**Describe the bug**

There's a bug in `NettyMessagingService` where on error we will keep track of open futures but not clear them. This originally came from https://github.com/camunda-cloud/zeebe/issues/6654, specifically this comment: https://github.com/camunda-cloud/zeebe/issues/6654#issuecomment-807136250

The original issue mentions a direct memory issue but I couldn't reproduce it, so I'll keep that issue open for now, but we can already fix the heap memory leak.

**To Reproduce**

See the original issue https://github.com/camunda-cloud/zeebe/issues/6654#issuecomment-807136250

**Expected behavior**

We clear futures out of the openFutures list once they have been completed/aren't necessary anymore.

**Environment:**
 - OS: 
 - Zeebe Version: 0.26+ (incl. 1.0)
 - Configuration:

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

Fixes a memory leak in the messaging service where connections opened with keep alive created a long living opened future which was never removed if the connection/channel initially failed to connect.

This was tested manually as there isn't an easy way to test for an absence of memory leak; please refer to the associated PR for more.

I ran the following test and measured the difference in VisualVM - every time, at the end of the test, I also performed GC multiple times and verified that it had run (via VisualVM as well):

```java
  // regression test https://github.com/camunda-cloud/zeebe/issues/6878
  @Test
  public void shouldNotLeakOpenFutures() throws InterruptedException {
    // given
    final var nonExistentAddress = Address.from(1);
    final var runtime = Runtime.getRuntime();

    // when
    runtime.gc();
    final long initialFreeMemory = runtime.freeMemory();
    for (int i = 0; i < 500_000; i++) {
      netty1.sendAndReceive(nonExistentAddress, "something or other", new byte[0], true);
    }

    // then
    Thread.sleep(50_000);
  }
```

> NOTE: this test was added to `NettyMessageServiceTest.java`

The results without the fix:

netty-mem-leak

The results with the fix:

netty-mem-leak-fix

## Related issues

closes #6878

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/NettyMessagingService.java [hunk 1]
```diff
@@ -438,11 +440,19 @@ private void initNioTransport() {
...
+                                  openFutures.remove(future);
+                                });
                           }
                         });
               } else {
-                executor.execute(() -> future.completeExceptionally(channelError));
+                executor.execute(
+                    () -> {
+                      future.completeExceptionally(channelError);
+                      openFutures.remove(future);
```
Comment (Zelldon) [inline atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/NettyMessagingService.java:454]: So this was the actual bug right? That it was not removed if we have a channelError?

**Review by ChrisKujawa — APPROVED** [review]
Looks good thanks :+1:

File: atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/NettyMessagingService.java (hunk 1 above)
Comment (npepinpe) [reply atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/NettyMessagingService.java:454]: Yes - also that it was removed from the list before being completed, but that had much less consequences.

**Comment by npepinpe** [comment]
bors retry

**Pull request merged**